File: app/product/internal/service/product/product.go
package product

import (
	"context"

	v1 "github.com/tmjwjx/supermarket/api/product/v1"
	bizproduct "github.com/tmjwjx/supermarket/app/product/internal/biz/product"
	"github.com/tmjwjx/supermarket/app/product/internal/service/principal"
)

type ProductService struct {
	v1.UnimplementedProductServiceServer
	uc *bizproduct.ProductUsecase
}

func NewProductService(uc *bizproduct.ProductUsecase) *ProductService {
	return &ProductService{uc: uc}
}

func (s *ProductService) CreateProduct(ctx context.Context, req *v1.CreateProductRequest) (*v1.CreateProductResponse, error) {
	if _, err := principal.RequireCatalogAdmin(ctx); err != nil {
		return nil, err
	}
	skus := make([]bizproduct.Sku, 0, len(req.GetSkus()))
	for _, sku := range req.GetSkus() {
		skus = append(skus, bizproduct.Sku{
			ID:          sku.GetId(),
			SpecsJSON:   sku.GetSpecsJson(),
			Price:       sku.GetPrice(),
			MarketPrice: sku.GetMarketPrice(),
			Image:       sku.GetImage(),
			Barcode:     sku.GetBarcode(),
			Enabled:     sku.GetEnabled(),
		})
	}
	created, err := s.uc.Create(ctx, &bizproduct.Product{
		CategoryID: req.GetCategoryId(),
		BrandID:    req.GetBrandId(),
		Name:       req.GetName(),
		Subtitle:   req.GetSubtitle(),
		Keywords:   req.GetKeywords(),
		Unit:       req.GetUnit(),
		DetailHTML: req.GetDetailHtml(),
		WeightGram: req.GetWeightGram(),
		Images:     append([]string(nil), req.GetImages()...),
		Skus:       skus,
		Params:     toParams(req.GetParams()),
	})
	if err != nil {
		return nil, err
	}
	return &v1.CreateProductResponse{Product: toProduct(created, true)}, nil
}

func (s *ProductService) PublishProduct(ctx context.Context, req *v1.PublishProductRequest) (*v1.PublishProductResponse, error) {
	op, err := principal.RequireCatalogAdmin(ctx)
	if err != nil {
		return nil, err
	}
	p, err := s.uc.Publish(ctx, req.GetId(), op)
	if err != nil {
		return nil, err
	}
	return &v1.PublishProductResponse{Product: toProduct(p, false)}, nil
}

func (s *ProductService) UnpublishProduct(ctx context.Context, req *v1.UnpublishProductRequest) (*v1.UnpublishProductResponse, error) {
	op, err := principal.RequireCatalogAdmin(ctx)
	if err != nil {
		return nil, err
	}
	p, err := s.uc.Unpublish(ctx, req.GetId(), op)
	if err != nil {
		return nil, err
	}
	return &v1.UnpublishProductResponse{Product: toProduct(p, false)}, nil
}

func (s *ProductService) ListProducts(ctx context.Context, req *v1.ListProductsRequest) (*v1.ListProductsResponse, error) {
	cards, next, err := s.uc.List(ctx, bizproduct.ListInput{
		CategoryID: req.GetCategoryId(),
		BrandID:    req.GetBrandId(),
		MinPrice:   req.GetMinPrice(),
		MaxPrice:   req.GetMaxPrice(),
		OrderBy:    req.GetOrderBy(),
		PageSize:   req.GetPageSize(),
		PageToken:  req.GetPageToken(),
		OnSaleOnly: true,
	})
	if err != nil {
		return nil, err
	}
	return &v1.ListProductsResponse{Products: toCards(cards), NextPageToken: next}, nil
}

func (s *ProductService) GetProduct(ctx context.Context, req *v1.GetProductRequest) (*v1.GetProductResponse, error) {
	p, err := s.uc.Get(ctx, req.GetId(), principal.OptionalUserID(ctx))
	if err != nil {
		return nil, err
	}
	out := toProduct(p, false)
	for _, opt := range bizproduct.SpecOptions(p.Skus) {
		out.SpecOptions = append(out.SpecOptions, &v1.SpecOption{Name: opt.Name, Values: opt.Values})
	}
	return &v1.GetProductResponse{Product: out}, nil
}

func (s *ProductService) GetProductDetail(ctx context.Context, req *v1.GetProductDetailRequest) (*v1.GetProductDetailResponse, error) {
	html, err := s.uc.Detail(ctx, req.GetId())
	if err != nil {
		return nil, err
	}
	return &v1.GetProductDetailResponse{DetailHtml: html}, nil
}

func (s *ProductService) SearchProducts(ctx context.Context, req *v1.SearchProductsRequest) (*v1.SearchProductsResponse, error) {
	cards, next, err := s.uc.Search(ctx, req.GetQ(), req.GetOrderBy(), req.GetPageToken(), req.GetPageSize())
	if err != nil {
		return nil, err
	}
	return &v1.SearchProductsResponse{Products: toCards(cards), NextPageToken: next}, nil
}

func (s *ProductService) BatchGetSkus(ctx context.Context, req *v1.BatchGetSkusRequest) (*v1.BatchGetSkusResponse, error) {
	rows, err := s.uc.BatchSkus(ctx, req.GetSkuIds())
	if err != nil {
		return nil, err
	}
	return &v1.BatchGetSkusResponse{Skus: toSnapshots(rows)}, nil
}

func (s *ProductService) CheckSellable(ctx context.Context, req *v1.CheckSellableRequest) (*v1.CheckSellableResponse, error) {
	rows, err := s.uc.CheckSellable(ctx, req.GetSkuIds())
	if err != nil {
		return nil, err
	}
	return &v1.CheckSellableResponse{Skus: toSnapshots(rows)}, nil
}

func (s *ProductService) IncreaseSales(ctx context.Context, req *v1.IncreaseSalesRequest) (*v1.IncreaseSalesResponse, error) {
	if err := s.uc.AddSales(ctx, "", req.GetOrderId(), req.GetProductId(), req.GetCount()); err != nil {
		return nil, err
	}
	return &v1.IncreaseSalesResponse{}, nil
}

func (s *ProductService) AdminListProducts(ctx context.Context, req *v1.AdminListProductsRequest) (*v1.AdminListProductsResponse, error) {
	if _, err := principal.RequireCatalogAdmin(ctx); err != nil {
		return nil, err
	}
	cards, next, err := s.uc.List(ctx, bizproduct.ListInput{
		PageSize:   req.GetPageSize(),
		PageToken:  req.GetPageToken(),
		OnSaleOnly: false,
	})
	if err != nil {
		return nil, err
	}
	return &v1.AdminListProductsResponse{Products: toCards(cards), NextPageToken: next}, nil
}

func toProduct(p *bizproduct.Product, withDetail bool) *v1.Product {
	if p == nil {
		return nil
	}
	out := &v1.Product{
		Id:         p.ID,
		CategoryId: p.CategoryID,
		BrandId:    p.BrandID,
		Name:       p.Name,
		Subtitle:   p.Subtitle,
		Keywords:   p.Keywords,
		MainImage:  p.MainImage,
		Unit:       p.Unit,
		WeightGram: p.WeightGram,
		Status:     p.Status,
		MinPrice:   p.MinPrice,
		MaxPrice:   p.MaxPrice,
		SalesCount: p.Sales,
		Images:     append([]string(nil), p.Images...),
	}
	if withDetail {
		out.DetailHtml = p.DetailHTML
	}
	for _, param := range p.Params {
		out.Params = append(out.Params, &v1.ProductParam{Name: param.Name, Value: param.Value})
	}
	for _, sku := range p.Skus {
		out.Skus = append(out.Skus, &v1.Sku{
			Id:          sku.ID,
			ProductId:   sku.ProductID,
			SpecsJson:   sku.SpecsJSON,
			Price:       sku.Price,
			MarketPrice: sku.MarketPrice,
			Image:       sku.Image,
			Barcode:     sku.Barcode,
			Enabled:     sku.Enabled,
		})
	}
	return out
}

func toCards(rows []bizproduct.Card) []*v1.ProductCard {
	out := make([]*v1.ProductCard, 0, len(rows))
	for _, row := range rows {
		out = append(out, &v1.ProductCard{
			Id:          row.ID,
			Name:        row.Name,
			MainImage:   row.MainImage,
			MinPrice:    row.MinPrice,
			MarketPrice: row.MarketPrice,
			SalesCount:  row.Sales,
			Status:      row.Status,
		})
	}
	return out
}

func toSnapshots(rows []bizproduct.SkuView) []*v1.SkuSnapshot {
	out := make([]*v1.SkuSnapshot, 0, len(rows))
	for _, row := range rows {
		out = append(out, &v1.SkuSnapshot{
			SkuId:       row.SkuID,
			ProductId:   row.ProductID,
			ProductName: row.Name,
			SpecsJson:   row.SpecsJSON,
			Price:       row.Price,
			Image:       row.Image,
			Sellable:    row.Sellable,
		})
	}
	return out
}

func toParams(rows []*v1.ProductParam) []bizproduct.Param {
	out := make([]bizproduct.Param, 0, len(rows))
	for _, row := range rows {
		if row == nil {
			continue
		}
		out = append(out, bizproduct.Param{Name: row.GetName(), Value: row.GetValue()})
	}
	return out
}

File: frontend/web/src/api/product.ts
import { send } from './client.ts'
import { asRecord, readList, readNumber, readString, unwrap } from './read.ts'

export const statusOnSale = 5
export const statusOff = 6

export type ProductCard = {
  id: string
  name: string
  image: string
  minPrice: number
  marketPrice: number
  sales: number
}

export type Sku = {
  id: string
  specs: string
  price: number
  marketPrice: number
  image: string
  enabled: boolean
}

export type SpecOption = {
  name: string
  values: string[]
}

export type Param = {
  name: string
  value: string
}

export type Product = {
  id: string
  name: string
  subtitle: string
  image: string
  status: number
  detailHtml: string
  skus: Sku[]
  specOptions: SpecOption[]
  params: Param[]
}

export type ProductPage = {
  products: ProductCard[]
  nextPageToken: string
}

export type ListQuery = {
  categoryId?: string
  brandId?: string
  minPrice?: number
  maxPrice?: number
  orderBy?: string
  pageSize?: number
  pageToken?: string
}

export function listProducts(query: ListQuery = {}) {
  const params = new URLSearchParams()
  if (query.categoryId) params.set('category_id', query.categoryId)
  if (query.brandId) params.set('brand_id', query.brandId)
  if (query.minPrice && query.minPrice > 0) params.set('min_price', String(query.minPrice))
  if (query.maxPrice && query.maxPrice > 0) params.set('max_price', String(query.maxPrice))
  if (query.orderBy) params.set('order_by', query.orderBy)
  if (query.pageSize) params.set('page_size', String(query.pageSize))
  if (query.pageToken) params.set('page_token', query.pageToken)
  const text = params.toString()
  return send(`/v1/products${text ? `?${text}` : ''}`, {})
}

export function searchProducts(q: string, pageToken = '', pageSize = 0, orderBy = '') {
  const params = new URLSearchParams({ q })
  if (orderBy) params.set('order_by', orderBy)
  if (pageSize) params.set('page_size', String(pageSize))
  if (pageToken) params.set('page_token', pageToken)
  return send(`/v1/products:search?${params.toString()}`, {})
}

export function getProduct(id: string) {
  return send(`/v1/products/${encodeURIComponent(id)}`, {})
}

export function getProductDetail(id: string) {
  return send(`/v1/products/${encodeURIComponent(id)}/detail`, {})
}

export function readProductDetail(body: unknown): string {
  const record = asRecord(body)
  if (!record) return typeof body === 'string' ? body : ''
  return readString(record, 'detail_html', 'detailHtml')
}

export function readProductCards(body: unknown): ProductCard[] {
  return readList(body, 'products').flatMap((row) => {
    const card = readCard(row)
    return card ? [card] : []
  })
}

export function readProductPage(body: unknown): ProductPage {
  const record = asRecord(body) ?? {}
  return {
    products: readProductCards(body),
    nextPageToken: readString(record, 'next_page_token', 'nextPageToken'),
  }
}

export function readProduct(body: unknown): Product | null {
  const product = unwrap(body, 'product')
  if (!product) return null
  const id = readString(product, 'id')
  if (!id) return null
  const skus = Array.isArray(product.skus)
    ? product.skus.flatMap((item) => {
        const row = asRecord(item)
        if (!row) return []
        const skuId = readString(row, 'id')
        if (!skuId) return []
        const enabled = row.enabled === undefined ? true : row.enabled === true
        return [{
          id: skuId,
          specs: readString(row, 'specs_json', 'specsJson'),
          price: readNumber(row, 'price'),
          marketPrice: readNumber(row, 'market_price', 'marketPrice'),
          image: readString(row, 'image'),
          enabled,
        }]
      })
    : []
  const specOptions = readList(product, 'spec_options').concat(readList(product, 'specOptions')).flatMap((row) => {
    const name = readString(row, 'name')
    const values = Array.isArray(row.values) ? row.values.filter((v): v is string => typeof v === 'string') : []
    return name && values.length ? [{ name, values }] : []
  })
  const params = readList(product, 'params').flatMap((row) => {
    const name = readString(row, 'name')
    return name ? [{ name, value: readString(row, 'value') }] : []
  })
  return {
    id,
    name: readString(product, 'name'),
    subtitle: readString(product, 'subtitle'),
    image: readString(product, 'main_image', 'mainImage'),
    status: readNumber(product, 'status'),
    detailHtml: readString(product, 'detail_html', 'detailHtml'),
    skus,
    specOptions,
    params,
  }
}

function readCard(row: Record<string, unknown>): ProductCard | null {
  const id = readString(row, 'id')
  if (!id) return null
  return {
    id,
    name: readString(row, 'name'),
    image: readString(row, 'main_image', 'mainImage'),
    minPrice: readNumber(row, 'min_price', 'minPrice'),
    marketPrice: readNumber(row, 'market_price', 'marketPrice'),
    sales: readNumber(row, 'sales_count', 'salesCount'),
  }
}
